### PR TITLE
Validate saved stats in CUDA batch norm backward

### LINKED ISSUE #189831: [CUDA] illegal memory read on batch_norm_backward_kernel
### 🐛 Describe the bug

The CUDA native_batch_norm_backward path launches over the channel dimension even when the reshaped input has zero elements and never checks that save_invstd is nonempty, so the kernel reads it through a null pointer.

```python
import torch

torch.cuda.init()
gpu = torch.device('cuda')

input = torch.full((4, 2, 0, 3), 1.0, device=gpu)
grad_out = torch.zeros(0, device=gpu)
weight = torch.full((2,), float('nan'), device=gpu)
running_mean = torch.full((2,), 1.0, device=gpu)
save_mean = torch.full((2,), 1e25, device=gpu)
save_invstd = torch.zeros(0, device=gpu)
torch.ops.aten.native_batch_norm_backward(
    grad_out, input, weight, running_mean, None,
    save_mean, save_invstd, True, float('-inf'),
    [True, False, False],
)
torch.cuda.synchronize()
```

```
========= COMPUTE-SANITIZER
========= Invalid __global__ read of size 4 bytes
=========     at void at::native::batch_norm_backward_kernel<float, float, float, int>(torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, const T1, (unsigned long)2, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)3, T4>, const T1, (unsigned long)3, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, const T1, (unsigned long)2, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)3, T4>, const T1, (unsigned long)3, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, T1, (unsigned long)2, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)3, T4>, T1, (unsigned long)3, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, T2, (unsigned long)0, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)1, T4>, T2, (unsigned long)1, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, T2, (unsigned long)0, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)1, T4>, T2, (unsigned long)1, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, const T2, (unsigned long)0, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)1, T4>, const T2, (unsigned long)1, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, const T2, (unsigned long)0, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)1, T4>, const T2, (unsigned long)1, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, const T2, (unsigned long)0, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)1, T4>, const T2, (unsigned long)1, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, const T3, (unsigned long)0, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)1, T4>, const T3, (unsigned long)1, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, const T3, (unsigned long)0, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)1, T4>, const T3, (unsigned long)1, torch::headeronly::DefaultPtrTraits, T4>, bool, T3)+0x90
=========     by thread (0,1,0) in block (0,0,0)
=========     Address 0x0 is out of bounds
=========     and is 140166624706560 bytes before the nearest allocation at 0x7f7b15e00000 of size 2097152 bytes
```

### Versions

Collecting environment information...

/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  104
On-line CPU(s) list:                     0-103
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Gold 5320 CPU @ 2.20GHz
CPU family:                              6
Model:                                   106
Thread(s) per core:                      2
Core(s) per socket:                      26
Socket(s):                               2
Stepping:                                6
CPU(s) scaling MHz:                      39%
CPU max MHz:                             3400.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4400.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               2.4 MiB (52 instances)
L1i cache:                               1.6 MiB (52 instances)
L2 cache:                                65 MiB (52 instances)
L3 cache:                                78 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-25,52-77
NUMA node1 CPU(s):                       26-51,78-103
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.13.0+cu126
[pip3] triton==3.7.1
[conda] Could not collect

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia

### PR DESCRIPTION
**Pull request by Claudio999222**
Fixes #189831

## Summary

`native_batch_norm_backward` could launch the CUDA backward kernel with
invalid saved statistics.

In training mode, the kernel reads one value per input channel from
`save_mean` and `save_invstd`. If either tensor does not contain exactly
one element per input channel, the kernel may perform an out-of-bounds
device read.

In the reported case, an empty `save_invstd` caused an illegal CUDA
memory access through a null pointer.

This change validates `save_mean` and `save_invstd` on the host before
creating the packed tensor accessors and launching the CUDA kernel.

Invalid saved statistics now raise a controlled `RuntimeError` instead
of causing a CUDA device memory fault. The execution path for valid
inputs remains unchanged.

## Test plan

The change was tested using both the dedicated regression test and a
locally rebuilt PyTorch installation containing the proposed fix.

### Regression test

Added a CUDA regression test covering empty `save_mean` and
`save_invstd` tensors.

Command:

```text
python test/test_nn.py -k native_batch_norm_backward_rejects_invalid_saved_stats

Result:

Ran 2 tests in 0.158s

OK (skipped=1)

The CUDA test passed successfully. One generated non-applicable test
variant was skipped.

Locally rebuilt PyTorch

PyTorch was rebuilt locally from source with the proposed change.

The original failure case was then executed against the rebuilt library.
The invalid input now raises the expected controlled error:

Expected save_invstd to contain 2 elements, but got 0

This confirms that the rebuilt native CUDA implementation contains and
executes the new host-side validation.

NVIDIA Compute Sanitizer

The original failure case was also executed against the locally rebuilt
PyTorch library using NVIDIA Compute Sanitizer.

Result:

PASS: Expected save_invstd to contain 2 elements, but got 0
ERROR SUMMARY: 0 errors

No invalid global reads, illegal CUDA memory accesses, or other device
memory errors were detected.

### CODE REVIEW

**Comment by Claudio999222** [comment]
@eqy @syed-ahmed @Aidyn-A  
Could I have some feedback please, if I need to change something I'll proceed.

Greetings C.

**Comment by eqy** [comment]
The original issue seems to be caused with an extremely contrived construction that doesn't seem to reflect any real use-cases and we are attempting to remedy it with a number of additional checks on the host side. Is there a measurable performance impact when doing this?

**Comment by Claudio999222** [comment]
> The original issue seems to be caused with an extremely contrived construction that doesn't seem to reflect any real use-cases and we are attempting to remedy it with a number of additional checks on the host side. Is there a measurable performance impact when doing this?

Thanks @eqy for raising this concern.

I benchmarked the exact parent commit against the commit containing the proposed validation, using the same build configuration, CUDA 12.6, and an NVIDIA RTX 4070 Laptop GPU.

The benchmark directly invoked `aten.native_batch_norm_backward` with valid saved statistics and covered:

- `(4, 2, 0, 3)`  original zero-sized shape
- `(2, 2, 1, 1)`  very small workload
- `(32, 64, 28, 28)`  representative workload
- `(16, 256, 14, 14)` representative higher-channel workload

Each case was measured repeatedly using `torch.utils.benchmark.Timer`.

The measurements showed noticeable run-to-run variability, including variation in the unchanged baseline itself. However, the patched version was not consistently slower than the baseline for either the small or representative workloads.
Therefore, no consistent and measurable performance regression was observed in these tests.
The added checks only inspect host-side tensor metadata through `defined()` and `numel()`. They do not access device memory or introduce a CUDA synchronization.
I can provide the benchmark script and complete raw measurements if that would be useful.

**Comment by Claudio999222** [comment]
@eqy
Before I start implementing a fix, I would like to confirm whether this is considered a valid real-world correctness problem, or whether it depends on a highly contrived input that is unlikely to occur through normal PyTorch usage.

I recently worked on a similar CUDA issue where the proposed fix introduced additional host-side validation, and the practical relevance of the reproducer became an important part of the review.

Would a fix for #189670 be useful and aligned with the expected behavior, or would you prefer a different direction?

Thanks.

File: test/test_nn.py [hunk 1]
```diff
@@ -7882,6 +7882,50 @@ def _test_LayerNorm_cuda_half(self, device):
         output.sum().backward()
         self.assertEqualTypeString(output, input)
 
+    @onlyCUDA
```
Comment (malfet) [inline test/test_nn.py]: Why onyCUDA?
```suggestion
```

**Comment by malfet** [comment]
@eqy I think extra metadata checks are fine, but I don't like the `onlyCUDA` nature of the fix

**Review by malfet — CHANGES_REQUESTED** [review]
This shape check must not be cuda specific, it must work for all backends, but your current implementation crashes on CPU (and MPS as well I suspect)

Also, when you writing a test, think if it's possible to cause by calling backward more organically, rather than calling specific ATEN op

If this happens only when one calls ATen and check is already present on the Python level, then python one must be removed

**Comment by Claudio999222** [comment]
> This shape check must not be cuda specific, it must work for all backends, but your current implementation crashes on CPU (and MPS as well I suspect)
> 
> Also, when you writing a test, think if it's possible to cause by calling backward more organically, rather than calling specific ATEN op
> 
> If this happens only when one calls ATen and check is already present on the Python level, then python one must be removed

Thanks for the review and feedback! I'll address the requested changes and push the updates shortly.
